### DISCUSSION TITLE
Support for supra connect

### DISCUSSION THREAD

**Discussion by hyonle** [BimmerGestalt/AAIdrive#636] [General]
Hi, would like to ask. If the team is considering adding support for supra connect.

Thank you.

**Comment by hufman** [comment]
Supra Connect does connect to the car over the BMW Apps protocol, so it's possible. However, the only authentication cert that has been found only allows access to the car data. AAIdrive's apps all use certs that are locked to BMW or MINI brands, and so it can not add any apps to the Supra.
AAIdrive does have an [15-month-old branch of code](https://github.com/BimmerGestalt/AAIdrive/tree/j29) that tries to use this Supra cert to connect to the car and show data, and enable the use of data-only addons. If you'd be willing to test, I would like to stabilize and merge this branch!

**Reply by hyonle** [reply]
How can I help? What is required.

**Reply by hufman** [reply]
I updated the branch, try running [this test build](https://bimmergestalt.s3.amazonaws.com/aaidrive/builds/androidautoidrive-latest-j29-nomap-sentry-release.apk) and let me know how it looks! It should detect that you have Supra Connect installed, it should connect to the car (untested) with a disclaimer saying most functionality doesn't work, and it should show data in the Car Information tab.

**Reply by hyonle** [reply]
tried it out. it does detects the supra connect . however it crashes when connecting.

here is a screen capture of the process.

https://www.dropbox.com/sh/sgrc02gnz2hzs75/AADkYBWPzmNQNIpYkhCZHIATa?dl=0

**Reply by hufman** [reply]
Ah thank you very much! As part of merging in the updated code, I missed a Supra-specific cert requirement which had moved to a different file in the intervening months. I sincerely appreciate you enabling crash reporting, that helped me immediately identify the problem!

Please update to the [latest APK](https://bimmergestalt.s3.amazonaws.com/aaidrive/builds/androidautoidrive-latest-j29-nomap-sentry-release.apk) and try again, thank you very much!

**Reply by hyonle** [reply]
the latest APK doesn't crash now :)

it looks like its all connecting, all the information shows up on AAidrive

- however i was unable to get the album art to show up on idrive via blue tooth audio
- have tired, deleting and re-pairing the phone, however there is no app options that shows up on idrive


screen capture and screen shots here.
https://www.dropbox.com/sh/sgrc02gnz2hzs75/AADkYBWPzmNQNIpYkhCZHIATa?dl=0

**Reply by hufman** [reply]
That's great to hear! Thank you so much for testing!

Unfortunately there is no way currently to add new car apps to Supra, without BMW issuing more certs for the J29 brand. The only functionality possible is retrieving the Car Information tab and any car data add-ons. The coverart feature is a part of the Spotify car app, which was only released for BMW and Mini.

I will need to do some minor tweaks (disabling the Navigation tab, for example) before merging it. Thank you again for testing!

**Comment by hufman** [comment]
Thank you so much for testing! I merged support in 39cd7c4
I'm sorry additional functionality can't be added to your Supra! If you find the car data useful, would you be willing to contribute an app translation in your language?

**Reply by hyonle** [reply]
english is my primary language. sorry i can't help in this aspect :)

**Reply by hufman** [reply]
Thank you anyways for testing and confirming that it works, finally letting me merge this branch of code :)

**Comment by Sq00ky** [comment]
Hi there @hufman
I have a '22 Supra and was wondering if there were any updates or if there's anything that I could do to help further the development for Supra support.

**Reply by hufman** [reply]
Thank you for reaching out!
The car only allows certs to log in that have a matching brand. All of the car app certs that have been found only include the BMW and MINI brands, except for one special cert that allows car data access from the J29 brand. Without finding any other J29 certs that enable access for adding car apps, AAIdrive can't add any more functionality for the Supra.

**Comment by hufman** [comment]
@hyonle @Sq00ky Do you still have your Supra vehicles, and are you comfortable with using BimmerCode?
It seems that IDrive will ignore the brand requirement if you code `ENTWICKLER_MENUE` to aktiv, would you like to try that? I believe AAIdrive may successfully connect with the BMW-brand cert and then connect the car apps like normal.
I'd appreciate any `adb logcat -d` files of your attempts!

**Reply by Sq00ky** [reply]
Hi @hufman,
Here's two separate log files post-coding. One is from directly after - the other is a shutdown/disconnect bluetooth/start the car/reconnect to bluetooth/open AAIDrive & browse a few settings..
[Uploading supra-ENTWICKLER_MENUE-enabled.zip…](https://github.com/user-attachments/files/16489370/supra-ENTWICKLER_MENUE-enabled.zip) If you have any questions or would like me to try anything else please let me know. I apologize for any background app noise (i.e. Discord & others).
Please note that I'll be out of town for DEFCON starting tomorrow, so any other trials may have to wait if not done tonight :(

**Reply by hufman** [reply]
Oh, nifty! Thank you so much!
I see from your FRESH log the following sequence of lines:
```
	Line 243137: 08-04 20:16:53.480  8762  9977 I CarProber: Probing detected a HMI type J29 ID6L
	Line 243297: 08-04 20:16:53.871  8762  9977 I CarProber: Probing detected a HMI type J29 ID6L
	Line 243436: 08-04 20:16:54.460  8762  9977 I CarProber: Probing detected a HMI type J29 ID6L
```
Examining the [car detection logic](https://github.com/BimmerGestalt/AAIdrive/blob/12de9790196f43cf424e950f0976d863bd810b53/app/src/main/java/me/hufman/androidautoidrive/CarProber.kt#L130-L172), it successfully logged in with the BMW cert and the Mini cert but skipped those loops because the HMI Type didn't match. When it got to the J29 cert, it just accepted the "j29" brand, which disables most of the connection logic in MainService.
Let me spin up a branch quick that handles BMW Cert with the J29 HMI!

**Reply by hufman** [reply]
@Sq00ky Please try [this build](https://bimmergestalt.s3.amazonaws.com/aaidrive/builds/androidautoidrive-latest-j29_as_bmw-mapbox-sentry-release.apk), which was [built](https://github.com/BimmerGestalt/AAIdrive/actions/runs/10240656269) from [this branch](https://github.com/BimmerGestalt/AAIdrive/tree/j29_as_bmw)

**Reply by Sq00ky** [reply]
![20240804_225438](https://github.com/user-attachments/assets/51207d84-6fc8-45f5-aae0-e7200d17f73e)
![20240804_225824](https://github.com/user-attachments/assets/2a8aa067-63f8-456e-b6c7-a1200e42079a)
![20240804_225223](https://github.com/user-attachments/assets/d821d08f-5c91-472c-98f5-03a354f97f21)
Bingo :)

**Reply by Sq00ky** [reply]
Updated log set if you want to take a peak - Let me know if there's any issues. All seems to be working. There are a few things here and there, like the "Detailed Vehicle Info" application has no name associated with it in the apps menu and the navigation menu seems to be wonky. will follow up with pictures in a few, but honestly isn't an issue because of the built in navigation functions in the car.
[Supra-ENTWICKLER_MENUE-post-j29.zip](https://github.com/user-attachments/files/16490208/Supra-ENTWICKLER_MENUE-post-j29.zip)
![20240804_231114](https://github.com/user-attachments/assets/d599c89e-ae26-4460-9678-b21a07849391)
![20240804_230714](https://github.com/user-attachments/assets/8c955333-6322-4a65-b5d9-328990b17b3c)

**Reply by hufman** [reply]
Whoa that's so cool! Excellent taste in music too!
The unlabeled Car Information icon is a known limitation. If you solder some UART and unlock SSH access and edit a .so file, there's a way around it (#801) but I haven't tried it myself.
The map dimensions are [defined here](https://github.com/BimmerGestalt/IDriveConnectKit/blob/main/core/src/commonMain/kotlin/io/bimmergestalt/idriveconnectkit/RHMIDimensions.kt) heuristically based on the HMI Type, and so it's falling through to a generic guess in your car. We can tune that if desired later, I have an ancient branch that shows custom setters to help tune the dimensions. The same library is used by the Screen Mirroring addon, so it might be worth it.

Thank you for helping, that's so exciting! I'll add some instructions to the app for other Supra owners to enjoy it!

**Reply by Sq00ky** [reply]
I'll have to look into it one day - it's not the first time I've ripped apart my dash or pulled the head unit, but it'd definitely be the first time soldering headers for UART and CLI access, haha.
Great work as always - I'm sure the rest of the Supra community is going to be thrilled. Having a nice middle ground in between an MMI box is wonderful for everyone.
Super glad you found this and always happy to help :)

**Reply by hufman** [reply]
I rebased the [example-submodule branch](https://github.com/BimmerGestalt/AAIdrive/tree/example-submodules), which should build AAIdrive with editable libraries instead of compiled artifacts. Check out [the developer docs](https://bimmergestalt.github.io/AAIdrive/developers#developing-idriveconnectkit) for more information about building AAIdrive and developing the IDriveConnectKit library.

**Comment by venkiiiiiiii** [comment]
Hi, I'm based in Australia where there is no Supra Connect, or Toyota connect app available.
Is there anyway to get this working without them?

**Reply by hufman** [reply]
Have you tried installing the Supra Connect app from a different region? For example, [here's North America](https://apkpure.com/toyota-supra-connect/de.j29.connected.mobile20.na). I'm surprised there's no app entirely, that's how you would manage your subscriptions to enable weather and traffic in the car.
Even if you get it working, Australia doesn't allow BMW to show any car app while driving, so AAIdrive doesn't provide much benefit. I haven't heard of any way around this.

**Reply by venkiiiiiiii** [reply]
Yeah unfortunately in Australia we don't have any connected services available for the Supra. 
I already have an MMI box for AA but figured this would be a nice thing to have...

**Comment by aprayiii** [comment]
Woohoo, texts in my Supra! Thank you for this app!

**Comment by aprayiii** [comment]
Currently not working because the latest SupraConnect update broke the app and it will not start. No SupraConnect means no AAIDrive.

**Reply by Sq00ky** [reply]
I just checked as this thread made me wonder if it still works - **for me it does**, I have a '22, latest version (5.7.0) of Supra Connect & AAIDrive (1.4.3). Running the Pixel 9 Pro on the latest OS/security patch. Attached is my log: [adb.zip](https://github.com/user-attachments/files/22061668/adb.zip).

**If you'd like to help**, enable USB Debugging on your Android device, it should be able to be found in Developer Settings. Afterwards, grab a laptop, follow the install instructions for ADB here:
https://www.xda-developers.com/install-adb-windows-macos-linux/

After it's installed, make sure you don't have any important apps open/running. Connect your phone to your PC. You should see a message on your phone that says "let this computer perform debugging things" or something along those lines. Click yes. Then open a command prompt (cmd.exe/powershell.exe/terminal/whatever) and run an **adb logcat >> adb.log**. 

Open Supra Connect, open AAIDrive, connect to Bluetooth on the car, ensure the car has the apps feature/setting enabled. See if you can do whatever you originally did to trigger the "rooted phone" message.

Afterwards, press Control+C in the terminal, zip up the log file, and post it here or in the issue you made.

**Reply by aprayiii** [reply]
My problem was not with AAIDrive, it was that the latest Supra Connect app would not start on my phone. Actually it would not start on my last THREE phones (since I still have them all). Current phone and last phone are Pixels, prior two before those Samsungs. They all give an error that the phone is rooted. (None are.) I was able to get a downlevel Supra Connect APK and it starts just fine. And AAI Drive works perfectly with it.

**Reply by aprayiii** [reply]
I called the help number for Supra Connect, they gave me a different number to call at BMW, they tried to open a ticket for me and told me their system is telling them that Supra Connect will now only work on i-Phone and is no longer supported for any Android device.

**Reply by aprayiii** [reply]
FYI that they must have fixed it on the server side because the latest 5.7 version of the Supra Connect app now works again.

**Reply by aprayiii** [reply]
Scratch that. Reverse it. 5.7 of Supra Connect stopped working again.
